include/deal.II/base/discrete_time.h

```diff
@@ -78,6 +78,141 @@ DEAL_II_NAMESPACE_OPEN
+ *   middle of this stage, some variables have been updated to $t_{n+1}$ but
+ *   other variables still represent their value at $t_n$. Thus, we call this
+ *   the inconsistent stage, requiring that no post-processing output take
+ *   place in this stage. The state variables get updated one by one.
+ *   In general, the order of updating variables is arbitrary, but it should
+ *   happen in a reasonable order. For example, if some variable such as $x$
+ *   depends on the calculation of another variable such as $y$, then $y$ must
+ *   be updated before $x$ can be updated. The question arises whether time
+ *   should be incremented before updating state quantities. Multiple
+ *   possibilities exist, depending on the programmer preferences:
+ *   * Time is incremented before the rest of the updates. In this case, even
+ *     though time is incremented to $t_{n+1}$, not all variables are updated
+ *     yet. During this update phase, $dt$ equals the *previous* time step
+ *     size. *Previous* means that it is referring to the  $dt$ of the
+ *     `advance_time()` command that was performed previously. In the
+ *     following example code, we are assuming that `a` and `b` are two state
+ *     variables that need to be updated in this time step.
+ *     @code
+ *       time.advance_time();
+ *       new_a = update_a(a, b, time.get_previous_step_size());
+ *       b = update_b(a, b, time.get_previous_step_size());
+ *       a = new_a;
+ *     @endcode
+ *   * Time is incremented from $t_n$ to $t_{n+1}$ after all variables have
+ *     already been updated for $t_{n+1}$. During the update stage, $dt$ is
+ *     denoted as the *next* time step size. *Next* means that $dt$ of the
+ *     step corresponds to the `advance_time()` command that will happen
+ *     subsequently.
+ *     @code
+ *       new_a = update_a(a, b, time.get_next_step_size());
+ *       b = update_b(a, b, time.get_next_step_size());
+ *       a = new_a;
+ *       time.advance_time();
+ *     @endcode
+ *   * Time is incremented in the middle of the other updates: In this case
+ *     $dt$ would correspond to *next* or *previous* depending of whether it
+ *     is used before or after the call to `advance_time()`.
+ *     @code
+ *       new_a = update_a(a, b, time.get_next_step_size());
+ *       time.advance_time();
+ *       b = update_b(a, b, time.get_previous_step_size());
+ *       a = new_a;
+ *     @endcode
+ *
+ * One thing to note is that, during the update phase, $dt$ is referred to
+ * either **next** or **previous** time step size, depending on whether the
+ * command `advance_time()` has been called yet. The notion of *current* time
+ * step size is ill-defined. In fact, in the update stage the definition of
+ * every variable depends on whether it has been updated yet or not, hence the
+ * name **the inconsistent stage**.
+ *
+ * The following code snippet shows the code sections for the snapshot stage
```

Comment:
I modified the final code example to use the style of our tutorials based on the file discrete_time_03.cc that you provided. Thank you

I will explain the `for loop` situation and the `while loop` alternative in a separate comment.

Reply:
Thank you for doing this, and sorry that it effectively overwrote what you'd originally 

Does your last sentence mean that you'll add an example using a `while loop` in a separate _commit_ (or follow up PR)? If so, that would be cool, as I think that it would then satisfy @bangerth's reservations in [this comment](https://github.com/dealii/dealii/pull/9748/files#r411701296). This is great stuff, but it's clearly not so easy to construct a setup that satisfies everyone. So thank you again for your patience and endeavouring to work through this with us.

Reply:
Nope, I've just spotted [the comment](https://github.com/dealii/dealii/pull/9748#issuecomment-616073523) that you were referring to. It's been a long day...